## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Комментарий наставника
Выводы стали гораздо лучше (как и обработка пропусков).Ppivot table ты честно попробовал, пока не очень наглядно, но все приходит с опытом.
Поздравляю, ты сдал первый проект! :-)

## Комментарий наставника
Привет! Спасибо, что быстро сдал задание:) Далее в файле мои комментарии ты сможешь найти в ячейках markdown с пометкой "комментарий наставника". Пострайся учесть эти комментарии в ходе выполнения дальнейших проектов. 
Я пока не могу принять твою работу (хотя многие веди сделанны верно и выводы в целом хорошие, и даже оформление мне в основном нравится). Главные причины такие:

1) кое-где нет текстовых пояснений, хотя они должны быть;

2) Ты не используешь pivot table, а это прям очень-очень важно. Мне необходимо, что бы ты хотя бы дважды использовал именно его. Я понимаю, что ты и так хорошо оформляешь свои данные. Они понятны и на них приятно посмотреть. Но давай попробуем сделать их еще лучше, хорошо?

Я написала тебе еще много всяких пожеланий, но они по факту являются именно пожалениями, а не требованиями. Можешь выполнять, если хочешь, можешь оставить как есть - смотри сам по своим целям и нагрузке. 

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Вывод
Импортируем библиотеку "Pandas" для работы с данными. Читаем наш искомый файл data.csv и изучаем его общую информацию.
В таблице 21525 строк, в столбцах "days_employed" и "total_income" есть нулевые объекты, их количество одинаково.

### Шаг 2. Предобработка данных

### Обработка пропусков

Пропущенные значения есть в столбцах "days_employed" и "total_income", так как наша работа связана с исследованием зависимости возврата кредита в срок от наличия детей и семейного положения то не будем удалять строки с пропущенными значениями, а заменим NaN на 0.
Пропущенные значения могли появиться из-за неверного заполнения таблицы сотрудниками.

In [2]:
data['days_employed']=data['days_employed'].fillna(0)
data['total_income']=data['total_income'].fillna(0)
data['total_income'].head(15)

0     253875.639453
1     112080.014102
2     145885.952297
3     267628.550329
4     158616.077870
5     255763.565419
6     240525.971920
7     135823.934197
8      95856.832424
9     144425.938277
10    113943.491460
11     77069.234271
12         0.000000
13    130458.228857
14    165127.911772
Name: total_income, dtype: float64

### Вывод

Методом info() проверяем данные и видим, что нулевых объектов не осталось.

## Комментарий наставника
Пожалуй, в данном случае "замена на 0" не является самым лучшим решением. Но это точно не самое плохое решение. Какие варианты еще возможны?

## Ответ на комментарий
Возможна замена на медианные значения, причем значения каждой группы людей, к которым относится человек. Смотрим по образованию, по должности и семейному положению.
Заменим доход в месяц на медианное значение соответствующей группы людей. Количество отработанных дней оставим 0, потому что не влияет на выводы в данной работе.

In [3]:
def function(row):
    try:
        if row['total_income']==0:
            return data[(data['education_id']==row['education_id']) & (data['income_type']==row['income_type']) &(data['family_status_id']==row['family_status_id'])]['total_income'].median()
        else: return row['total_income']
    except:
        print('Ошибка где-то')
data['total_income'] = data.apply(function, axis=1)
print(data['total_income'].head(15))
data.info()

0     253875.639453
1     112080.014102
2     145885.952297
3     267628.550329
4     158616.077870
5     255763.565419
6     240525.971920
7     135823.934197
8      95856.832424
9     144425.938277
10    113943.491460
11     77069.234271
12    104903.875260
13    130458.228857
14    165127.911772
Name: total_income, dtype: float64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2

## Комментарий наставника
Реешение стало существенно более корректным)

### Замена типа данных

В столбцах "days_employed" и "total_income" стоит тип данных - float64. Что вероятно получилось от деления общего дохода, на месяцы.
Заменим на int, так как отработанных дней не может быть дробное количество как и рублей.

In [4]:
data['days_employed']=data['days_employed'].astype('int')
data['total_income']=data['total_income'].astype('int')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


### Вывод

Методом info() проверяем данные и видим, что заменен вещественный тип данных на целочисленный.
Для изменения типа данных используем метод astype().

## Комментарий наставника
Здесь все правильно)

### Обработка дубликатов

Для начала найдем, сколько дубликатов есть. Приведем строковые значения к нижнему регистру.

In [5]:
data['education'] = data['education'].str.lower()
data['family_status'] = data['family_status'].str.lower()
data['gender'] = data['gender'].str.lower()
data['income_type'] = data['income_type'].str.lower()
data.duplicated().sum()

71

Удалим повторяющиеся значения.

In [6]:
data = data.drop_duplicates().reset_index(drop = True)
data.duplicated().sum()

0

### Вывод

### Исправлено
Так как дубликатов всего 71 значений, что незначительно по сравнению с 21525, то удаляем повторяющиеся строки методом drop_duplicates() с восстановлением индексов reset_index()

## Комментарий наставника
Здесь все правильно сделанно и объясненно. Тллько дубликатов 71, а не 17 - но это совсем не критичная опечатка.

### Лемматизация

In [7]:
from pymystem3 import Mystem
m = Mystem()
lemmas = []
def lemma_data(row):
    lemma = ' '.join(m.lemmatize(row))
    return lemma
data['lemma'] = data['purpose'].apply(lemma_data)
data[['lemma','purpose']].head(5)

lemma                     purpose
0               покупка   жилье \n               покупка жилья
1     приобретение   автомобиль \n     приобретение автомобиля
2               покупка   жилье \n               покупка жилья
3  дополнительный   образование \n  дополнительное образование
4             сыграть   свадьба \n             сыграть свадьбу

### Вывод

Для обработки данных введем еще один столбец с списком лемматизированных слов, путем применения нашей функции 'lemma_data', которая возвращает лемматизированный вид слов, к столбцу "purpose". Видим, что отработало корректно.

## Комментарий наставника
Тут проблема - слетел вывод. Будь внимательней, хорошо?

### Категоризация данных

Категоризируем данные по типу наличия детей. Будут категории - "Нет детей", "Один ребенок", "Два ребенка", "Многодетные". Так как в таблице есть данные с -1 ребенком в семье, то удалим эти строки.

In [8]:
data = data.loc[data['children'] != -1]
def childrens(row):
    if row == 1:
        return 'Один ребенок'
    if row == 2:
        return 'Два ребенка'
    if row > 2:
        return 'Многодетные'
    if row == 0:
        return 'Нет детей'
data_childrens = data.loc[:,['children','debt']]
data_childrens['childrens'] = data_childrens['children'].apply(childrens)
data_childrens['childrens'].head()

0    Один ребенок
1    Один ребенок
2       Нет детей
3     Многодетные
4       Нет детей
Name: childrens, dtype: object

Категоризируем данные по уровням дохода. "До 50 тысяч в месяц" "От 50 до 100" и "От 100 до 200 тысяч" и "От 200 до 300 тысяч" и "От 300 тысяч". Убираем все строки с доходом отрицательным и равным 0.

In [9]:
data = data.loc[data['total_income'] >0]
def finance(row):
    if row <= 50000:
        return 'До 50 тысяч в месяц'
    elif 50000<row<=100000:
        return 'От 50 до 100'
    elif 150000>row >=100000:
        return 'От 100 до 150 тысяч'
    elif 200000>row >=150000:
        return 'От 150 до 200 тысяч'
    elif 250000>row >=200000:
        return 'От 200 до 250 тысяч'
    elif row >=250000:
        return 'Больше 250 тысяч'
data_finance = data.loc[:,['total_income','debt']]
data_finance['total_income_new'] = data_finance['total_income'].apply(finance)
data_finance['total_income_new'].value_counts().head(15)

От 100 до 150 тысяч    7117
От 150 до 200 тысяч    4771
От 50 до 100           4087
Больше 250 тысяч       2807
От 200 до 250 тысяч    2251
До 50 тысяч в месяц     372
Name: total_income_new, dtype: int64

Категоризируем данные по целям кредита. Просмотрим и проанализируем список уникальных значений и выделим из него основные цели кредита. Это будут - автомобиль, свадьба, недвижимость, образование.

In [10]:
from collections import Counter
#print(Counter(data['lemma']))
def purposing(row):
    if 'автомобиль' in row:
        return 'Приобретение автомобиля'
    elif 'свадьба' in row:
        return 'На свадьбу'
    elif 'недвижимость' in row:
        return 'Приобретение недвижимости'
    elif 'жилье' in row:
        return 'Приобретение недвижимости'
    elif 'образование' in row:
        return 'На образование'
data_purpose = data.loc[:,['lemma','debt']]
data_purpose['purpose'] = data_purpose['lemma'].apply(purposing)

### Вывод

Категоризируем данные по типу наличия детей. Будут категории - "Нет детей", "Один ребенок", "Два ребенка", "Многодетные". Так как в таблице есть данные с -1 ребенком в семье, то удалим эти строки.¶

Категоризируем данные по уровням дохода. "До 50 тысяч в месяц" "От 50 до 100" и "От 100 до 150 тысяч" и "От 150 до 200 тысяч" и "От 200 до 300 тысяч" и "От 300 тысяч". Убираем все строки с доходом отрицательным и равным 0.¶


Категоризируем данные по целям кредита. Просмотрим и проанализируем список уникальных значений и выделим из него основные цели кредита. Это будут - автомобиль, свадьба, недвижимость, образование.

## Комментарий наставника
Тут тоже нет вывода (точнее, он улетел выше), а он мне необходим. Пожалуйста, пиши их в нужном месте, а то я могу и проглядать (все мы люди, а люди ошибаются). 

На мой взгляд, не стоит делить так грубо "есть дети/нет детей". Бизнесу интересно еще и есть ли связь между "качеством" заемщика и количеством его детей. Но твой вариант тоже имеет право на жизнь.

Щаг разбиения дохода ты выбрал весьма не плохо, хотя я бы продолжила разбиение и после достиженяи дохода в 100 тыс.

Разбиение по целям кредита сделанно просто образцово-показательно. Цели выделены верно.

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [11]:
#print(data_childrens.groupby(['children', 'debt'])['debt'].count())

children  debt
0         0       13028
          1        1063
1         0        4364
          1         444
2         0        1858
          1         194
3         0         303
          1          27
4         0          37
          1           4
5         0           9
20        0          68
          1           8
Name: debt, dtype: int64


In [12]:
#print('Процент невернувших долг людей с детьми = {:%}'.format(677/6639))
#print('Процент невернувших долг людей без детей = {:%}'.format(1063/13028))

Процент невернувших долг людей с детьми = 10.197319%
Процент невернувших долг людей без детей = 8.159349%


In [13]:

data_pivot = data_childrens.pivot_table(index=['debt'], columns='childrens', values='children', aggfunc='count')
print(data_pivot.head(10))
print('Процент невернувших долг людей без детей = {:%}'.format(data_pivot.loc[1,'Нет детей']/data_pivot.loc[0,'Нет детей']))
print('Процент невернувших долг людей с одним ребенком = {:%}'.format(data_pivot.loc[1,'Один ребенок']/data_pivot.loc[0,'Один ребенок']))
print('Процент невернувших долг людей с двумя детьми = {:%}'.format(data_pivot.loc[1,'Два ребенка']/data_pivot.loc[0,'Два ребенка']))
print('Процент невернувших долг многодетных = {:%}'.format(data_pivot.loc[1,'Многодетные']/data_pivot.loc[0,'Многодетные']))


childrens  Два ребенка  Многодетные  Нет детей  Один ребенок
debt                                                        
0                 1858          417      13028          4364
1                  194           39       1063           444
Процент невернувших долг людей без детей = 8.159349%
Процент невернувших долг людей с одним ребенком = 10.174152%
Процент невернувших долг людей с двумя детьми = 10.441335%
Процент невернувших долг многодетных = 9.352518%


### Старый вывод
Нельзя сказать, что есть зависимость, между наличием детей и возвратом кредита в срок. Скорее наоборот, люди необремененные чадом, возвращают кредиты на 2% чаще.

### Новый вывод
Многодетные семьи возвращают долги исправнее на 1%, чем люди с одним ребенком или двумя (10.17% и 10.44% должников соотвественно). Люди необремененные чадом, всё еще возвращают кредиты на 2% чаще.

## Комментарий наставника
Вывод верен. Сюда бы еще информацию о том, как количество детей влияет на возврат долга.

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [14]:
print(data.groupby(['family_status', 'debt'])['debt'].count())

family_status          debt
в разводе              0        1105
                       1          85
вдовец / вдова         0         892
                       1          63
гражданский брак       0        3758
                       1         388
женат / замужем        0       11379
                       1         930
не женат / не замужем  0        2531
                       1         274
Name: debt, dtype: int64


In [15]:
print('Процент невернувших долг людей в разводе = {:%}'.format(85/1105))
print('Процент невернувших долг людей потерявших супруга = {:%}'.format(63/892))
print('Процент невернувших долг людей в гражданском браке = {:%}'.format(388/3758))
print('Процент невернувших долг людей в браке = {:%}'.format(930/11379))
print('Процент невернувших долг холостых людей = {:%}'.format(274/2531))

Процент невернувших долг людей в разводе = 7.692308%
Процент невернувших долг людей потерявших супруга = 7.062780%
Процент невернувших долг людей в гражданском браке = 10.324641%
Процент невернувших долг людей в браке = 8.172950%
Процент невернувших долг холостых людей = 10.825761%


### Вывод
Холостые люди самые безответственные и невозвращают кредит в 11% случаев. Тогда как потерявшие супруга и люди в разводе склонны возвращать кредиты в 7% случаев.

## Комментарий наставника
Вывод корректен.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [16]:
#print(data_finance.groupby(['total_income_new', 'debt'])['debt'].count())

total_income_new     debt
Больше 250 тысяч     0       2613
                     1        194
До 50 тысяч в месяц  0        349
                     1         23
От 100 до 150 тысяч  0       6488
                     1        629
От 150 до 200 тысяч  0       4372
                     1        399
От 200 до 250 тысяч  0       2087
                     1        164
От 50 до 100         0       3756
                     1        331
Name: debt, dtype: int64


In [17]:
#print('Процент невернувших долг людей с доходом большее 100 тысяч = {:%}'.format(726/8439))
#print('Процент невернувших долг людей с доходом до 50 тысяч = {:%}'.format(23/348))
#print('Процент невернувших долг людей с доходом от 50 до 100 тысяч = {:%}'.format(821/8950))

Процент невернувших долг людей с доходом большее 100 тысяч = 8.602915%
Процент невернувших долг людей с доходом до 50 тысяч = 6.609195%
Процент невернувших долг людей с доходом от 50 до 100 тысяч = 9.173184%


In [18]:
data_pivot = data_finance.pivot_table(index=['debt'], columns='total_income_new', values='total_income', aggfunc='count')
print(data_pivot.head(10))
print('Процент невернувших долг людей с доходом до 50 тысяч = {:%}'.format(data_pivot.loc[1,'До 50 тысяч в месяц']/data_pivot.loc[0,'До 50 тысяч в месяц']))
print('Процент невернувших долг людей с доходом от 50 до 100 тысяч = {:%}'.format(data_pivot.loc[1,'От 50 до 100']/data_pivot.loc[0,'От 50 до 100']))
print('Процент невернувших долг людей с доходом от 100 до 150 тысяч = {:%}'.format(data_pivot.loc[1,'От 100 до 150 тысяч']/data_pivot.loc[0,'От 100 до 150 тысяч']))
print('Процент невернувших долг людей с доходом от 150 до 200 тысяч = {:%}'.format(data_pivot.loc[1,'От 150 до 200 тысяч']/data_pivot.loc[0,'От 150 до 200 тысяч']))
print('Процент невернувших долг людей с доходом от 200 до 250 тысяч = {:%}'.format(data_pivot.loc[1,'От 200 до 250 тысяч']/data_pivot.loc[0,'От 200 до 250 тысяч']))
print('Процент невернувших долг людей c доходом больше 250 тысяч = {:%}'.format(data_pivot.loc[1,'Больше 250 тысяч']/data_pivot.loc[0,'Больше 250 тысяч']))

total_income_new  Больше 250 тысяч  До 50 тысяч в месяц  От 100 до 150 тысяч  \
debt                                                                           
0                             2613                  349                 6488   
1                              194                   23                  629   

total_income_new  От 150 до 200 тысяч  От 200 до 250 тысяч  От 50 до 100  
debt                                                                      
0                                4372                 2087          3756  
1                                 399                  164           331  
Процент невернувших долг людей с доходом до 50 тысяч = 6.590258%
Процент невернувших долг людей с доходом от 50 до 100 тысяч = 8.812567%
Процент невернувших долг людей с доходом от 100 до 150 тысяч = 9.694821%
Процент невернувших долг людей с доходом от 150 до 200 тысяч = 9.126258%
Процент невернувших долг людей с доходом от 200 до 250 тысяч = 7.858170%
Процент невернувших дол

### Старый вывод
Самые ответственные люди с наименьшим доходом, до 50 тысяч. Люди с доходом от 50 до 100 не возвращают долги в 9% случаев. Возможна корреляция с суммами кредитов, так как богатые люди берут большие кредиты.

### Новый вывод
С увеличением доходов до 150 тысяч количество должников растет. После 150 тысяч в месяц падает. Самые ответственные люди - с доходом до 50 тысяч и больше 250 тысяч в месяц. 

## Комментарий наставника
Вот тут тебя сильно подвело, что ты не вывел следующие разбиения - от 100 до 150 тыс, от 150 до 200 тыс и далее. Вывод мог получится иным.

- Как разные цели кредита влияют на его возврат в срок?

### Вывод

In [19]:
print(data_purpose.groupby(['purpose', 'debt'])['debt'].count())

purpose                    debt
На образование             0        3632
                           1         370
На свадьбу                 0        2136
                           1         186
Приобретение автомобиля    0        3893
                           1         402
Приобретение недвижимости  0       10004
                           1         782
Name: debt, dtype: int64


In [20]:
print('Процент невернувших долг людей бравших кредит на образование = {:%}'.format(370/3632))
print('Процент невернувших долг людей бравших кредит на свадьбу = {:%}'.format(186/2136))
print('Процент невернувших долг людей бравших кредит на автомобиль = {:%}'.format(402/3893))
print('Процент невернувших долг людей бравших кредит на недвижимость = {:%}'.format(782/10004))

Процент невернувших долг людей бравших кредит на образование = 10.187225%
Процент невернувших долг людей бравших кредит на свадьбу = 8.707865%
Процент невернувших долг людей бравших кредит на автомобиль = 10.326227%
Процент невернувших долг людей бравших кредит на недвижимость = 7.816873%


## Вывод
Люди берущие заемные средства на приобретение недвижимости самые ответсвенные. Тогда как автомобилисты чаще просрачивают возврат денег.
##### Дополнение
Вместе с приобретающими автомобиль часто не возвращают догли люди берущие кредиты на образование.

## Комментарий наставника
Какая еще группа кредитов очень близка по своей рискованности к кредиту на автомобиль? Разница ведь почти не существенная (десятые доли процентов). 

### Шаг 4. Общий вывод

Меньше всего банкам стоит доверять человеку с ребенком, холостому, берущему кредит на автомобиль и имеющему доход от 100 до 150 тысяч.

Самая надежная категория - берущий средства на приобретение недвижимости, с доходом от 50 тысяч (либо от 250), в разводе или вдова/вдовец, без детей.

## Комментарий наставника
Вывод в целом верен. Он скорее поможет банку, чем нет. Но хорошо бы еще указать самых лучших клиентов (это тоже оооочень интересно). Задачка на "подумать, елси есть время" - будут ли пожелния к изменению занесения информации для сотрудников? Что нужно изменить в программе, какие ограниченяи стоит поставить?

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.